### PR TITLE
addrman: Use system time instead of adjusted network time

### PR DESCRIPTION
**Pull request by maflcko**
This changes addrman to use system time for address relay instead of the network adjusted time.

This is an improvement, because network time has multiple issues:

* It is non-monotonic, even if the system time is monotonic.
* It may be wrong, even if the system time is correct.
* It may be wrong, if the system time is wrong. For example, when the node has limited number of connections (`4`), or the system time is wrong by too much (more than +-70 minutes), or the system time only got wrong after timedata collected more than half of the entries while the time was correct, ...)

This may slightly degrade addr relay for nodes where timedata successfully adjusted the time. Addr relay can already deal with minor offsets of up to 10 minutes. Offsets larger than this should still allow addr relay and not result in a DoS.

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK

**Review by mzumsande — COMMENTED** [review]
Concept ACK

I think that [this line](https://github.com/bitcoin/bitcoin/blob/3740cdd12551fab2cc9ef12afff8680c34086343/src/net_processing.cpp#L2927) in `net_processing` should be changed too, otherwise we'd still use adjusted time for addr relay.

> However, addr relay can already deal with minor offsets of up to 60 minutes

Where does the 60 minute number come from? 

I can think of the following relevant times:
When receiving addr messages ([relevant code](https://github.com/bitcoin/bitcoin/blob/3740cdd12551fab2cc9ef12afff8680c34086343/src/net_processing.cpp#L2964-L2976)):
* If `nTime` is >10 minutes in the future relative to our time, we backdate `nTime` to (OurTime - 5days) for addrman and don't relay it further. 
* If `nTime` is >10 minutes in the past, we don't relay it further (but accept it to addrman).

This means that addr gossip relay is relatively sensitive to what we think is the right time. If our time is off by more than 10 minutes, we will most likely not take part in it and not forward addresses received from others.
That would not prevent us from accepting addrs to addrman / finding outbound peers. But we wouldn't help the network and our own advertised address would be off by 10 minutes and not relayed by others as well, so we'd likely have a harder time getting inbound peers.

For GETADDR, we treat addresses as `Terrible` and don't include them in a GETADDR response if:
* If `nTime` is more than 10 minutes in the future (unlikely to happen, we would have backdated nTime if it was in the future when receiving it in the first place)
* If `nTime` is older than 30 days

So I think that GETADDR responses are not very sensitive at the time scales of AdjustedTime (minutes), especially since they are only calculated only every ~24h (GETADDR caching).

@naumenkogs mentioned [here](https://github.com/bitcoin/bitcoin/pull/23807#issuecomment-998621857) the possibility of fingerprinting:
>Also, making it further from mainstream (not using adjusted time) might add more fingerprinting capabilities by looking at those timestamps, for nodes with broken clocks (and mildly for the rest, as a consequence)

I think that's an interesting question, there are probably two aspects:
* Active, targeted fingerprinting by trying to adjust the time of a peer is made impossible by this.
* Passive fingerprinting of nodes that happen to be a few seconds off (which could have been corrected by the network before) could become easier.

**Comment by naumenkogs** [comment]
Concept ACK.

--------

>This means that addr gossip relay is relatively sensitive to what we think is the right time. If our time is off by more than 10 minutes, we will most likely not take part in it and not forward addresses received from others.
That would not prevent us from accepting addrs to addrman / finding outbound peers. But we wouldn't help the network and our own advertised address would be off by 10 minutes and not relayed by others as well, so we'd likely have a harder time getting inbound peers.

I was thinking more about our own peer selection in this case, especially deprioritizing in the context of `10+ minutes in the future get 5 days of penalty`.

I don't think this PR significantly worsens any existing behavior or opens any new attack surface.

For the "natural" behavior, it doesn't matter much because addr relay freshness has a lag of hours, not minutes (maybe an issue with immediate self-announcements, but those are already connected).
For the attacks, an attacker could always manipulate the addr timestamps how they wanted, and this PR doesn't make it easier.

---------

Re fingerprinting, I think that would be cool to investigate even for the current master.

----------

Otherwise I agree with the summary of @mzumsande.

**Comment by mzumsande** [comment]
> I was [thinking more](https://github.com/bitcoin/bitcoin/pull/23807#discussion_r824470943) about our own peer selection in this case, especially deprioritizing in the context of `10+ minutes in the future get 5 days of penalty`.

If I'm not missing something, we don't deprioritize based on `nTime`: In `GetChance()` which is called from `Addrman::Select_()`, there is logic to deprioritize based on (possibly failed) earlier attempts to connect, but I didn't find any influence of the  timestamp on our own peer selection.

**Comment by MarcoFalke** [comment]
> I think that [this line](https://github.com/bitcoin/bitcoin/blob/3740cdd12551fab2cc9ef12afff8680c34086343/src/net_processing.cpp#L2927) in net_processing should be changed too, otherwise we'd still use adjusted time for addr relay.

Good catch, fixed.

> Where does the 60 minute number come from? 

Thanks, fixed up. I think I incorrectly interpreted 600 as 60 minutes, whereas it is 10 minutes.

> Passive fingerprinting of nodes that happen to be a few seconds off (which could have been corrected by the network before) could become easier.

Bitcoin Core will already send it's system time in the version message. See commit 14ba286556faad794f288ef38493c540382897cb, so I don't think fingerprinting gets worse by this?

File: src/net.cpp [hunk 1]
```diff
@@ -2080,7 +2080,7 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
 
         addrman.ResolveCollisions();
 
-        int64_t nANow = GetAdjustedTime();
+        int64_t nANow = GetTime();
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/net.cpp]: Thx, done in a later commit

**Comment by MarcoFalke** [comment]
An alternative idea to this pull (not sure if serious):

Since we "know" the time offset for each peer, at least incoming ones (and outgoing ones after commit 14ba286556faad794f288ef38493c540382897cb), an alternative might also be to adjust the timestamp of each addr message on receipt by the peer-specific offset. This would get rid of adjusted time, but presumably limit the slight negative effects of our clock being wrong on addr relay?

**Comment by laanwj** [comment]
> An alternative idea to this pull (not sure if serious):

It's a intriguing idea, but I don't think it's worth it. Peers that have time offsets large enough for this to matter are probably unreliable anyway, and if our own time is wrong for even a short while, when it gets corrected the entire addrman db has wrong timestamps. It's also quite some extra complexity to test.

**Comment by MarcoFalke** [comment]
Right, this is probably not worth to look into due to the complexity and questionable benefit.

I'll rebase the changes here once https://github.com/bitcoin/bitcoin/pull/24697 is merged.

**Comment by fanquake** [comment]
Concept ACK

**Review by jonatack — COMMENTED** [review]
Concept ACK on the refactorings, modulo "addrman: Use system time instead of adjusted network time" that I think it would be wiser to hold off on for the time being (pun intended) (edit: or maybe code it in a way that is easily reversible).

**Comment by MarcoFalke** [comment]
rebased

**Comment by dergoegge** [comment]
Code review ACK fadd8b2676f6d68ec87189871461c9a6a6aa3cac

**Pull request merged**